simde/x86/avx512/roundscale.h

```diff
@@ -0,0 +1,40 @@
+#if !defined(SIMDE_X86_AVX512_ROUNDSCALE_H)
+#define SIMDE_X86_AVX512_ROUNDSCALE_H
+
+#include "types.h"
+#include "../sse.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+#if defined(SIMDE_X86_AVX512VL_NATIVE) && defined(SIMDE_X86_AVX512F_NATIVE)
+  #define simde_mm_roundscale_ps(a, imm8) _mm_roundscale_ps((a), (imm8))
+#else
+  SIMDE_FUNCTION_ATTRIBUTES
+  simde__m128
+  simde_mm_roundscale_ps (simde__m128 a, int imm8)
+      SIMDE_REQUIRE_CONSTANT_RANGE(imm8, 0, 255) {
+    simde__m128 temp, r, clear_sign;
+    int32_t
+      exponent1 = (127 + ((imm8 >> 4) & 15)) << 23,
+      exponent2 = (127 - ((imm8 >> 4) & 15)) << 23;
+
+    SIMDE_CONSTIFY_16_(simde_mm_round_ps, temp, simde_mm_setzero_ps(), imm8 & 15, simde_mm_mul_ps(a, simde_mm_set1_ps(*HEDLEY_REINTERPRET_CAST(SIMDE_FLOAT32_TYPE*, &exponent1))));
```

Comment:
Older GCC dies here due to dereferencing the type-punned pointer; see https://nemequ.semaphoreci.com/jobs/a75620d6-70ea-456d-ba79-a6740c911367

Instead, you should use `simde_memcpy`, which will copy the underlying data without aliasing:

```c
simde_float32 exponent1f;
simde_memcpy(&exponent1f, &exponent, sizeof(exponent1f));
```

Compilers are very good about getting rid of the memcpy, so this won't actually create any cost at runtime.

Reply:
Also, you should move the simde_mm_mul_ps call out of the CONSTIFY macro.  Everything in the arguments to the macro gets duplicated 16 times, so it's better to minimize that.

Finally, for the unused value, you should use `(HEDLEY_UNREACHABLE(), simde_mm_setzero_ps())`.  The `HEDLEY_UNREACHABLE` lets most compilers know that that code can never be reached, so it doesn't have to handle it.  The `simde_mm_setzero_ps()` part is only for compilers which `HEDLEY_UNREACHABLE()` can't be implemented for, which may emit a warning if you don't handle the case.